satellite/overlay/combinedcache.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+type addressInfo struct {
+	address   string
+	lastIP    string
+	transport pb.NodeTransport
+}
+
+type uptimeInfo struct {
+	isUp       bool
+	lastUptime time.Time
+	stats      *NodeStats
+}
+
+// CombinedCache is a simple caching mechanism for overlaycache updates. It
+// provdes methods to help reduce calls to UpdateAddress and UpdateTime, but can
+// be extended for other calls in the future.
+type CombinedCache struct {
+	DB
+	addressLock  sync.RWMutex
+	addressCache map[storj.NodeID]*addressInfo
+
+	uptimeLock          sync.RWMutex
+	uptimeCache         map[storj.NodeID]*uptimeInfo
+	uptimeFlushInterval time.Duration
+}
+
+// NewCombinedCache instantiates a new CombinedCache
+func NewCombinedCache(db DB, uptimeFlushInterval time.Duration) *CombinedCache {
+	return &CombinedCache{
+		DB:                  db,
+		addressCache:        make(map[storj.NodeID]*addressInfo),
+		uptimeCache:         make(map[storj.NodeID]*uptimeInfo),
+		uptimeFlushInterval: uptimeFlushInterval,
+	}
+}
+
+// UpdateUptime overrides the underlying db.UpdateUptime and provides a simple
+// caching layer to reduce calls to the underlying db.
+func (c *CombinedCache) UpdateUptime(ctx context.Context, nodeID storj.NodeID, isUp bool, lambda, weight, uptimeDQ float64) (stats *NodeStats, err error) {
+	// First check the internal cache. If it returns stats, use them.
+	stats = c.GetNodeStats(nodeID, isUp)
+	if stats != nil {
+		return stats, nil
+	}
+	stats, err = c.DB.UpdateUptime(ctx, nodeID, isUp, lambda, weight, uptimeDQ)
+	if err != nil {
+		return nil, err
+	}
+	// Refresh internal stats
+	c.SetNodeStats(nodeID, isUp, stats)
+	return stats, nil
+}
+
+// UpdateAddress overrides the underlying db.UpdateAddress and provides a simple
+// caching layer to reduce calls to the underlying db.
+func (c *CombinedCache) UpdateAddress(ctx context.Context, info *pb.Node, defaults NodeSelectionConfig) (err error) {
+	// Update internal cache and check if this call requires a db call
+	if !c.SetAndCompareAddress(info) {
+		return nil
+	}
+	return c.DB.UpdateAddress(ctx, info, defaults)
+}
+
+// SetAndCompareAddress returns true if the address should be updated in the
+// underlying db, and false if not. It also keeps the internal cache up-to-date.
+func (c *CombinedCache) SetAndCompareAddress(node *pb.Node) bool {
+	// There's nothing we can do with a nil node, or nil address, we're just
+	// gonna say don't update it
+	if node == nil {
+		return false
+	}
+
+	address := node.Address
+	if address == nil {
+		address = &pb.NodeAddress{}
+	}
+
+	c.addressLock.RLock()
+	cached, ok := c.addressCache[node.Id]
+	c.addressLock.RUnlock()
```

Comment:
This gap between RUnlock and Lock is another race, wherein another goroutine could change what's in the map and cause results to be applied out of order. If the race I mentioned above is deemed not a problem, then maybe this one isn't either.

Still, this one has an easier fix: just use Mutex instead of RWMutex, and you won't have to reacquire the lock in order to do a write. I know how attractive read-write mutexes are, believe me, but every time I use them it seems to end up being overcomplication or premature optimization. Read-write mutexes are quite a bit slower than plain mutexes—with a good implementation it may be on the order of 3x as long to do a `(*RWMutex).RLock()` than a `(*Mutex).Lock()`. And that is still fine if there are lots of readers and lots of lock contention and read locks tend to be held for a nontrivial amount of time, because the additional work allows many more readers to progress in parallel. But in this case, there is only a single map lookup going on with the read-lock held. I suspect that the map lookup might well take less time than the RLock or the RUnlock next to it. So if there is so much lock contention that an RWMutex is justified, then there is a very very large number of calls going on to RLock and RUnlock, and the cpu will end up taking a lot more total time in locking and unlocking than it does in moving the program forward.

Reply:
I believe this is also fixed in my newest implementation.

I think in this case the RW Mutex is worthwhile, since in the majority of cases the UpdateAddress calls should end up being returned from the cache, making an actual Lock call unnecessary. But let me know if you disagree after seeing the final code update.

Reply:
I probably haven’t explained very well. The RLock() operation itself is the slow thing. It has to perform more than one synchronization primitive under the covers in order to enforce the read/write semantics (I don’t know exactly which approach Go takes here, but I don’t know of any way to do it without multiple steps). Calling Lock() on a plain Mutex should be significantly faster. RWMutex generally probably isn’t worth it unless readers tend to hold their read locks for nontrivial amounts of time, so that it helps for them to be able to overlap ownership. Of course, I could be wrong about all of this and maybe Go will surprise me again, so if you have timing data that says RWMutex and RLock is better, then great! But I’m sufficiently sure about it that I think it’s worth checking.

Reply:
As we discussed in slack some, I ran some benchmark tests that suggested that, while RLock may be slightly slower than a regular Lock, the difference is a few nanoseconds. By keeping the RLock, we optimize the case (which I hope is the common case) where eventually we have cached most node addresses, and the cache (with an RLock) is able to concurrently handle almost all `UpdateAddress` calls.

Reply:
Random seagull flyby knowledge poop: in the case of no pending writers, an `RLock` is just an atomic add. If there is a pending writer, it’s just that add and a semaphore wait. It’s implemented by a counter of pending read locks that the `Lock` call atomically “negates” so that readers can semaphore.

The main performance problem with an rw-mutex comes from 1. too many write locks, 2. acquiring a read lock still is a write to a shared bit of memory, and so doesn’t scale very well to high core counts. It’s bottlenecked by processor cache level coordination.

Reply:
Well, the atomic add is still being used there as a synchronization primitive, with its own associated costs (at least one forced cache miss, possibly more, plus a memory fence, plus potential retries if it’s implemented with Intel CAS).

I’ve timed the Go implementation before, under moderate contention, and the 3x cost I’m used to from pthreads world was still about the same. Although, I don’t remember if that was on ARM or x86_64.